### ISSUE TITLE
[Phase 6] Visualization Suite

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/feynman-kac-pinn#7]
## Summary

Create rich scientific visualizations for random walk trajectories, PDE solutions, training dynamics, and error analysis using Plotly.js.

**Parent Issue:** #1
**Depends On:** #6 (Phase 5: React Frontend)

## Learning Context

Before coding, understand:
- **Plotly.js**: Interactive charting library with 3D support
- **react-plotly.js**: React wrapper for Plotly
- **Scientific Visualization**: Best practices for displaying mathematical data

## Files to Create

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `frontend/src/components/visualizations/RandomWalkViewer.tsx` | ~150 | Create | 3D random walk paths |
| `frontend/src/components/visualizations/SolutionSurface.tsx` | ~120 | Create | Solution heatmap |
| `frontend/src/components/visualizations/TrainingCurve.tsx` | ~80 | Create | Loss over time |
| `frontend/src/components/visualizations/ErrorAnalysis.tsx` | ~100 | Create | Error distribution |
| `frontend/src/components/visualizations/ConvergenceStudy.tsx` | ~100 | Create | MC convergence |
| `frontend/src/components/visualizations/index.ts` | ~20 | Create | Module exports |

### Task 1: Random Walk Visualization
- [ ] Display 3D scatter/line plot of walk paths
- [ ] Color paths by exit time
- [ ] Add animation controls for path progression
- [ ] Support dimension selection for high-D (show 3D slice)

### Task 2: Solution Surface
- [ ] 2D heatmap for solution values
- [ ] Interactive dimension selector for high-D
- [ ] Contour overlay option
- [ ] Side-by-side comparison with MC estimates

### Task 3: Training Curve
- [ ] Loss vs. training step
- [ ] Log-scale option
- [ ] Learning rate schedule overlay
- [ ] Interactive zoom and pan

### Task 4: Error Analysis
- [ ] Histogram of pointwise errors
- [ ] Spatial error heatmap
- [ ] Comparison with analytical solution (if available)
- [ ] Summary statistics panel

### Task 5: MC Convergence
- [ ] Variance vs. number of samples
- [ ] Overlay theoretical O(1/sqrt(N)) curve
- [ ] Confidence interval bands

## Code Skeleton

```tsx
// frontend/src/components/visualizations/RandomWalkViewer.tsx
import { useMemo } from 'react';
import Plot from 'react-plotly.js';
import { Data, Layout } from 'plotly.js';

interface Props {
  paths: number[][][];  // [n_paths][n_steps][dim]
  exitTimes: number[];
  showAnimation?: boolean;
  dimensions?: [number, number, number];  // Which 3 dims to show
}

export function RandomWalkViewer({
  paths,
  exitTimes,
  showAnimation = false,
  dimensions = [0, 1, 2],
}: Props) {
  const [dimX, dimY, dimZ] = dimensions;

  const traces = useMemo<Data[]>(() => {
    const maxExitTime = Math.max(...exitTimes);

    return paths.map((path, i) => ({
      type: 'scatter3d' as const,
      mode: 'lines',
      x: path.map(p => p[dimX]),
      y: path.map(p => p[dimY]),
      z: path.map(p => p[dimZ]),
      line: {
        color: exitTimes[i],
        colorscale: 'Viridis',
        cmin: 0,
        cmax: maxExitTime,
        width: 2,
      },
      opacity: 0.6,
      showlegend: false,
    }));
  }, [paths, exitTimes, dimX, dimY, dimZ]);

  const layout: Partial<Layout> = {
    title: 'Random Walk Trajectories',
    scene: {
      xaxis: { title: `Dimension ${dimX}` },
      yaxis: { title: `Dimension ${dimY}` },
      zaxis: { title: `Dimension ${dimZ}` },
    },
    margin: { l: 0, r: 0, b: 0, t: 40 },
  };

  return (
    <Plot
      data={traces}
      layout={layout}
      config={{ responsive: true }}
      style={{ width: '100%', height: '500px' }}
    />
  );
}
```

```tsx
// frontend/src/components/visualizations/SolutionSurface.tsx
import { useMemo } from 'react';
import Plot from 'react-plotly.js';

interface Props {
  xRange: [number, number];
  yRange: [number, number];
  values: number[][];  // 2D grid of solution values
  title?: string;
  showContours?: boolean;
}

export function SolutionSurface({
  xRange,
  yRange,
  values,
  title = 'PDE Solution',
  showContours = true,
}: Props) {
  const [xMin, xMax] = xRange;
  const [yMin, yMax] = yRange;

  const trace = useMemo(() => {
    const nX = values.length;
    const nY = values[0]?.length || 0;

    return {
      type: 'heatmap' as const,
      z: values,
      x: Array.from({ length: nX }, (_, i) => xMin + (i / (nX - 1)) * (xMax - xMin)),
      y: Array.from({ length: nY }, (_, j) => yMin + (j / (nY - 1)) * (yMax - yMin)),
      colorscale: 'RdBu',
      colorbar: { title: 'u(x,y)' },
    };
  }, [values, xMin, xMax, yMin, yMax]);

  const contourTrace = showContours ? {
    type: 'contour' as const,
    z: values,
    x: trace.x,
    y: trace.y,
    contours: { coloring: 'lines' },
    line: { color: 'black', width: 1 },
    showscale: false,
  } : null;

  return (
    <Plot
      data={contourTrace ? [trace, contourTrace] : [trace]}
      layout={{
        title,
        xaxis: { title: 'x' },
        yaxis: { title: 'y' },
        margin: { l: 60, r: 20, b: 60, t: 40 },
      }}
      config={{ responsive: true }}
      style={{ width: '100%', height: '400px' }}
    />
  );
}
```

```tsx
// frontend/src/components/visualizations/TrainingCurve.tsx
import Plot from 'react-plotly.js';

interface Props {
  history: { step: number; loss: number; learning_rate?: number }[];
  logScale?: boolean;
}

export function TrainingCurve({ history, logScale = true }: Props) {
  const lossTrace = {
    type: 'scatter' as const,
    mode: 'lines',
    x: history.map(h => h.step),
    y: history.map(h => h.loss),
    name: 'Loss',
    line: { color: '#3b82f6', width: 2 },
  };

  const lrTrace = history[0]?.learning_rate !== undefined ? {
    type: 'scatter' as const,
    mode: 'lines',
    x: history.map(h => h.step),
    y: history.map(h => h.learning_rate!),
    name: 'Learning Rate',
    yaxis: 'y2',
    line: { color: '#f59e0b', width: 1, dash: 'dot' },
  } : null;

  return (
    <Plot
      data={lrTrace ? [lossTrace, lrTrace] : [lossTrace]}
      layout={{
        title: 'Training Progress',
        xaxis: { title: 'Step' },
        yaxis: {
          title: 'Loss',
          type: logScale ? 'log' : 'linear',
        },
        yaxis2: {
          title: 'Learning Rate',
          overlaying: 'y',
          side: 'right',
          type: 'log',
        },
        legend: { x: 0.7, y: 0.95 },
        margin: { l: 60, r: 60, b: 50, t: 40 },
      }}
      config={{ responsive: true }}
      style={{ width: '100%', height: '350px' }}
    />
  );
}
```

```tsx
// frontend/src/components/visualizations/ErrorAnalysis.tsx
import Plot from 'react-plotly.js';

interface Props {
  predictions: number[];
  groundTruth: number[];
  coordinates?: number[][];  // For spatial visualization
}

export function ErrorAnalysis({ predictions, groundTruth, coordinates }: Props) {
  const errors = predictions.map((p, i) => p - groundTruth[i]);
  const absErrors = errors.map(Math.abs);

  const meanError = errors.reduce((a, b) => a + b, 0) / errors.length;
  const rmse = Math.sqrt(absErrors.map(e => e ** 2).reduce((a, b) => a + b, 0) / errors.length);
  const maxError = Math.max(...absErrors);

  const histogramTrace = {
    type: 'histogram' as const,
    x: errors,
    nbinsx: 50,
    marker: { color: '#3b82f6' },
    name: 'Error Distribution',
  };

  return (
    <div className="space-y-4">
      <div className="grid grid-cols-3 gap-4">
        <div className="p-4 bg-gray-50 rounded-lg text-center">
          <p className="text-sm text-gray-500">Mean Error</p>
          <p className="text-xl font-semibold">{meanError.toExponential(3)}</p>
        </div>
        <div className="p-4 bg-gray-50 rounded-lg text-center">
          <p className="text-sm text-gray-500">RMSE</p>
          <p className="text-xl font-semibold">{rmse.toExponential(3)}</p>
        </div>
        <div className="p-4 bg-gray-50 rounded-lg text-center">
          <p className="text-sm text-gray-500">Max Error</p>
          <p className="text-xl font-semibold">{maxError.toExponential(3)}</p>
        </div>
      </div>

      <Plot
        data={[histogramTrace]}
        layout={{
          title: 'Error Distribution',
          xaxis: { title: 'Prediction Error' },
          yaxis: { title: 'Count' },
          margin: { l: 60, r: 20, b: 50, t: 40 },
        }}
        config={{ responsive: true }}
        style={{ width: '100%', height: '300px' }}
      />
    </div>
  );
}
```

## Visualization Guidelines

1. **Color Schemes**: Use colorblind-friendly palettes (Viridis, RdBu)
2. **Interactivity**: Enable zoom, pan, hover tooltips by default
3. **Responsiveness**: All charts should resize with container
4. **Performance**: Limit number of points displayed (downsample if needed)
5. **Accessibility**: Include alt text and ARIA labels

## Verification Criteria

| Visualization | Test | Expected |
|--------------|------|----------|
| Random Walk | 100 paths render | < 1 second |
| Solution Surface | 100x100 grid | Smooth interaction |
| Training Curve | 5000 points | Responsive zoom |
| Error Analysis | Histogram + stats | Correct calculations |

## Definition of Done

- [ ] All 5 visualization types implemented
- [ ] Interactive controls work smoothly
- [ ] Export to PNG/SVG available
- [ ] Works on Chrome, Firefox, Safari
- [ ] Mobile touch gestures supported
